README.md
# netnrf 响应式框架
> 用于快速开发的响应式框架

> 演示一：<https://rf2.netnr.com>  
> 演示二：<https://free-rf.zme.ink>

### [更新日志](CHANGELOG.md)

### [文档说明](DOCUMENT.md)

### v3.x
- 前端采用 jQuery + Bootstrap + EasyUI + AceAdmin + fontAwesome
- 后端采用 .NET Core + EF + SQL（SQLServer、MySQL、PostgreSQL、SQLite）
    - 数据库脚本放置于 `wwwroot/scripts/`目录下
    - 数据库转换使用的工具：https://fishcodelib.com/DBMigration.htm
    - 全部采用 LINQ，跨数据库、避免SQL注入
- Visual Studio 2019 运行项目

### 项目结构
- Netnr.Core 类库（引用NuGet，源码地址：<https://github.com/netnr/Netnr.ClassLibrary>）
- Netnr.Data 数据访问、仓储（Scaffold-DbContext 命令自动生成）
- Netnr.Domain 实体（Scaffold-DbContext 命令自动生成）
- Netnr.Fast 常用方法
- Netnr.Func 应用
- Netnr.ResponseFramework Web站点

### 数据表
- 用户（SysUser）
- 角色、角色权限（SysRole）
- 菜单（SysMenu）
- 按钮（SysButton）
- 日志（SysLog）
- 字典（SysDictionary）
- 表配置（SysTableConfig）

### 功能
- 登录：系统账号登录
- 权限：角色权限，控制菜单及页面按钮
- 表格：动态配置标题、宽度、排序、对齐方式、格式化、冻结、点击排序等
- 表单：动态生成表单，自定义标题、排序、跨列、类型、必填等，支持多表单生成
- 查询：动态生成查询面板，自定义字段查询，以JSON格式表达查询条件
- 日志：访问日志记录
- 字典：通用的字典表
- 工具：数据库表信息展示，一键导出数据库表结构为Excel
- 任务：定时执行任务，清理临时目录、重置演示数据库
- 导出：公共导出Excel表，自定义查询主体，支持条件查询、列格式化，支持追加操作等
- 上传：通用的上传接口
- 接口：所有非页面请求规范化为接口，并用 swagger 生成可视化接口文档，<https://rf2.netnr.com/swagger>

### 使用说明
1. 创建表、写字段注释（方便生成表配置）
2. 生成表配置，可以用【工具箱】-【表管理】-【生成表配置】，也可以直接拷贝文件夹`wwwroot/scripts/table-config/`对应的`SQL`脚本运行
3. 修改表配置，表格，表单、查询，调整为需要展示的形式（标题、宽度、排序、输入类型、列格式化、必填、默认值等，根据业务拓展配置项）
4. 修改表配置，输入类型配置，需要配置下拉框、下拉树等，在`Common`控制器写方法，`url`源指向这个方法访问的地址
5. 修改表配置，列格式化配置，比如状态需要格式化为`启用`、`停用`，有常用公共的格式化方法，也可以配置自定义格式化方法`col_custom_字段小写`
6. 创建一个页面，菜单表添加此页面，配置操作按钮
7. 写表对应的查询、保存（新增/修改）、删除方法，参考【系统设置】里面的功能
8. 基于`z.js`封装的表格方法（API与EasyUI保持一致，看EasyUI文档即可），配置查询表的请求地址、表格类型、分页、复选等

### 截图

#### 列表 

![列表](https://static.netnr.com/2018/05/18/403ce7d002.png)

#### 新增、编辑、查看

![表单](https://static.netnr.com/2018/05/18/8d25d345b2.png)

#### 列表配置

![列表配置](https://static.netnr.com/2018/05/18/13da6572a3.png)

#### 表单配置

![表单配置](https://static.netnr.com/2018/05/18/0c98ee578c.png)

#### 角色权限配置（树）

![角色权限配置](https://static.netnr.com/2018/08/16/31a55cac78.png)

### 第三方文档API
- [EasyUI文档](https://ad.netnr.com/#EasyUI-1.5.2)
- [jQuery文档](https://ad.netnr.com/#jQuery-1.11.3)

### 附
- [联系（打赏）](https://ss.netnr.com/contact)
- [加入QQ群](http://qm.qq.com/cgi-bin/qm/qr?k=oLmAflGAIODgeYw9tImSvBVX1SK_warh)

### Source
- <https://github.com/netnr/netnrf>
- <https://gitee.com/netnr/netnrf>

CHANGELOG.md
# 更新日志

### [v3.1.2] - 2020-04-18
- 调整 组件CDN源
- 调整 升级引用包

### [v3.1.1] - 2019-12-22
- 调整 数据库上下文改为连接池托管
- 调整 IP地址查询 改为 `IP2Region` 组件

### [v3.1.0] - 2019-12-04
- 调整 升级为 `.NET Core 3.1`
- 添加 服务器信息 页面

### [v3.0.1] - 2019-11-13
- 调整 所有非页面请求规范化为接口，并用 swagger 生成可视化接口文档，<https://rf2.netnr.com/swagger>

### [v3.0.0] - 2019-09-29
- 调整 升级为 `.NET Core 3.0`

### [v2.2.10] - 2019-07-23
- 添加 `QueryWhere`针对`IEnumerable`条件查询的方法重载（支持部分查询关系符）
- 调整 表管理支持搜索（基于`IEnumerable`的查询，即内存搜索）
- 调整 日期时间组件，支持 **placeholder** 显示
- 修复 批处理切换方法`z.batchButtonSwitch`逻辑缺陷
- 添加 单据（主从表展示，如：采购订单，更多的组件支持须完善）
- 调整 表单组件传参调整，为兼容单据，即支持模态表单和页面切换单据的生成
- 调整 `z.js` 内部方法构造调整，解决了一些问题
- 调整 引入定时任务组件 **FluentScheduler**
- 添加 自动生成单表增删改查代码（Controller、View、JavaScript）

### [v2.2.9] - 2019-07-05
- 修复 未将定时任务的服务添加到容器，导致定时任务未执行

### [v2.2.8] - 2019-06-25
- 添加 `数据字典`
- 添加 表单嵌入富文本示例（基于 `CKEditor4`，图片支持粘贴、拖拽上传，大小限制 **2MB** ）
- 调整 z.Combo绑定，支持 **placeholder** 显示

### [v2.2.7] - 2019-06-23
- 修复 公共导出（js脚本问题）
- 添加 上传的示例（ajax上传，带上传真实进度条，表格显示格式化，查看文件，必填验证）
- 添加 批量操作示例（依赖 Z.EntityFramework.Extensions）

### [v2.2.6] - 2019-06-14
- 调整 输入输出实体对象名，为公共导出提供支持
- 调整 公共导出查询依赖查询，仅维护一处
- 调整 增删改统一返回格式，维护方法返回对象
- 调整 `appsettings.json` 配置文件设置键名 `TypeDB` 等于数据库类型
- 添加 `PostgreSQL`生成表配置、表信息查询的脚本

### [v2.2.5] - 2019-06-10
- 修复 在linux上导出、文件清理的路径问题
- 修复 内置参数可忽略日志记录的逻辑问题
- 修复 表单编辑，多选下拉列表值为空时赋值了一项空值的问题
- 添加 表配置新增项：表单`maxlength`属性
- 添加 表配置新增项：`查询条件关系符`配置，避免不支持的关系符查询出错
- 添加 Iframe页面`全屏`功能按钮
- 添加 低版本浏览器拦截提示页面
- 添加 下拉列表面板高度自适应
- 调整 DataGrid分页组件不显示调转页，启用快捷页码
- 调整 页面表单配置，拖拽排序的方式从节点交换改为节点插入
- 调整 窗口大小改变时从修改所有iframe标签高度改为仅修改当前iframe标签高度，切换时再触发
- 调整 为更好的自适应效果，iframe标签设置`scrolling="no"`，子页面有需要滚动条显示用`<div class="bodyscroll">`包装，参考桌面页

### [v2.2.4] - 2019-06-02
- 添加 角色添加复制角色权限功能
- 添加 `公共导出`，自定义查询、条件、格式化
- 添加 定时任务，自动重置数据库、清理临时目录
- 添加 表格示例，快捷查询与查询面板共存；注意：启用快捷查询的条件在查询面板中需关闭，避免冲突
- 添加 表格示例，`批量删除`功能示例
- 调整  `批处理`功能按钮优化，批处理状态功能按钮自动切换
- 调整 查询面板 默认关系 从  **等于**  改为  **包含**

### [v2.2.3] - 2019-04-26
- 调整 标签页添加控制菜单
- 调整 圆角调整为`1px`
- 调整 `EasyUI`默认引用`metro`主题
- 调整 静态资源引入由局部视图统一维护
- 添加 日志表添加IP所属的城市，离线查询IP地址，基于免费的`ipip.ipdb`
-  **据群发起投票结果，今后仅更新 `.NET Core` 版本，`.NET4` 停更** 

### [v2.2.2] - 2019-04-18
- 调整 日志分批写入

### [v2.2.1] - 2019-02-28
- 调整 导出 Excel的 `NPOITo` 类，支持 `.xls`、`.xlsx`导出
- 调整 表格查询面板生成模式，基于表格列配置生成，支持多表格查询面板
- 移除 快捷键 `Ctrl + Q` 打开查询面板（因存在多个查询面板）
- 添加 表格列标题快捷打开定位条件查询面板
- 调整 首次运行项目，自动创建数据库、调重置数据库（`scripts/table-reset` 目录需有对应的数据库脚本）
- 修复 datetime日期时间输入类型查询面板赋值存在的问题（没阻止冒泡，先结束编辑再赋值造成赋值失败）
- 调整 `Netnr.Core` 类库项目从NuGet引入，新增 `Netnr.Fast` 项目，具体看 https://github.com/netnr/Netnr.ClassLibrary
- 调整 支持多表单生成（页面的表单、表格配置默认第一个，暂不支持多表单、表格配置）

### [v2.2.0] - 2019-01-17
- 调整 Grid行数可选列表,可选200行
- 修复 多个Grid传参有干扰的问题,修复方式为先清空 `queryparams` 对象
- 调整 去除日期时间插件 `bootstrap-datetimepicker`  改为 `EasyUI` 提供曰期时间组件，减少两次请求
- 添加 Modal弹出窗设置 `autorefresh` 属性，每次打开 `iframe` 自动刷新
- 调整 Moda1弹出窗默认为全屏弹窗
- 调整 表单必填项提示改为红星
- 添加 中文语言包
- 修复 页面嵌套大于两层出现样式不一样的问题,修复方式为把 `parent` 对象改成 `top`
- 调整 `init.css` 样式包嵌入 `z.css`
- 调整 `easel-custom.css` 调整的样式嵌入 `z.css` ,页面直接引入原有包 `easyul.css`
- 调整 表配置隐藏添加系统级隐藏
- 调整 Grid行号宽度自适应
- 修复 切换用户菜单异常的问题，菜单缓存问题，去除了菜单缓存
- 添加 表管理导出 `Excel` 
- 调整 数据库表字段命名，减少关联查询出现重复列的一些问题
- 调整 演示项目数据库从 `MySQL` 迁移到 `SQLServer`,无其他原因，主要是自己用着顺手些
- 调整 查询面板默认关系符为 “等于”，支持快捷键 `Ctrl + Q` 打开查询面板、 `Enter` 确认搜索

### [v2.1.2] - 2018-11-09
- 增加 `.NET Framework 4.5`框架源代码，同步更新Core版本

### [v2.1.2] - 2018-11-01
- 修复 `DataGrid`执行`updateRow`更新有翻页时行号有问题（没设置pageNumber的原因）
- 调整 公共查询的一些操作

### [v2.1.2] - 2018-10-27
- 添加 `按钮管理`功能
- 添加 `菜单管理`功能
- 添加 `授权关联`功能（支持第三方登录）
- 添加 `z.Combo`方法添加清除值按钮支持
- 添加 `PostgreSQL`数据库支持
- 调整 `Linq`查询，可不传排序列，即默认排序
- 调整 选项卡右侧仅为刷新按钮

### [v2.1.1] - 2018-10-13
- 修复 方法`z.FindTreeNode`存在的问题
- 修复 方法`z.FormEdit`存在的问题
- 修复 `Linq`查询先分页后排序的问题
- 修复 系统操作日志，`IP`获取始终为`127.0.0.1`的问题，原因是`nginx`代理，需要判断代理环境

### [v2.1.0]　2018-10
- 添加 跨平台跨数据库支持，<https://github.com/aspnet/EntityFrameworkCore> ，已测试`SQLite`、`MySql`、`SQLServer`
- 生成实体依赖于`Scaffold-DbContext`命令 <https://www.netnr.com/gist/code/5283651389582691979>
- 调整 公共查询从SQL语句改为Linq
- 修复 `z.js`若干问题
- 添加 表管理工具，生成表配置、表字典
- 调整 演示项目调整，从`SQLServer`改为`MySql`，服务器迁移国外搬瓦工（Centos7、Nginx、MySql、CN2线路）
---
- 顺便说一下这段时间的体会，一直用`Windows`服务器，接触`Linux`后，认为`Linux`做服务器是挺好的，已打算全部迁移至`Linux`服务器。
  -  `Linux`服务器便宜，相对而言
  - 国外的`VPS`大多限流量不限速度（比如1G的带宽，每月1T的流量），当然线路也重要，不然延迟掉包严重
  - 不用备案、可以搭梯子翻墙
  - `Linux`开机占用`100`MB左右，跑个`dotnet`进程也才`300`MB左右
  - `.net framework`的项目也能跑哦，`mono` 、[jexus](https://www.jexus.org/)
  - `SQLServer`数据库已经有`Linux`版本，当然也可以改为`MySql`
  - 上手`Linux`服务器，一脸懵逼，什么都不知道不知从何下手，怎么办，你说怎么办，凉拌，网上有大把的教程，弄懂一个算一个；这里提供一些我用到的东西：
    - ssh命令连接服务器
    - 安装环境：`dotnet`、`nginx`、`ftp`、`mysql`、`frp`（微信开发）、`shadowsocks`（翻墙）
    - 学习`vi`编辑器，学习`dotnet`命令，学习`nginx`配置
    - 存的一些干货 <https://www.netnr.com/gist/user/1>

### [v2.0.0] - 2018-07
- 前端采用 jQuery + Bootstrap + EasyUI + AceAdmin
- 后端采用 Asp.Net Core + EF + SQLServer
- 全新重写 `z.js` 脚本包，与EasyUI提供的API高度保持一致（最大调整）
- 重写iframe选项卡
- 精简ace导航

src/Netnr.ResponseFramework/Netnr.ResponseFramework/wwwroot/js/tool/serverinfo.js
function loadOSinfo() {
    $.ajax({
        url: "/Tool/QueryServerInfo",
        dataType: 'json',
        success: function (data, status, xhr) {
            if (data.code != 200) {
                return false;
            }

            var arh = xhr.getAllResponseHeaders();

            data = data.data;

            var htm = [];
            htm.push('<table class="table table-bordered mt15">');

            arh.replace(/server: (.*)/, function () {
                htm.push('<tr>');
                htm.push('<td>服务</td>');
                htm.push('<td>' + arguments[1] + '</td>');
                htm.push('</tr>');
            })

            htm.push('<tr>');
            htm.push('<td>框架</td>');
            htm.push('<td>' + data.FrameworkDescription + '</td>');
            htm.push('</tr>');

            htm.push('<tr>');
            htm.push('<td>系统</td>');
            htm.push('<td>' + data.OSDescription + '</td>');
            htm.push('</tr>');

            htm.push('<tr>');
            htm.push('<td>内核</td>');
            htm.push('<td>' + data.OSVersion.VersionString + '</td>');
            htm.push('</tr>');

            htm.push('<tr>');
            htm.push('<td>CPU</td>');
            htm.push('<td>' + data.ProcessorName + ' ，' + data.ProcessorCount + ' Core</td>');
            htm.push('</tr>');

            htm.push('<tr>');
            htm.push('<td>内存</td>');
            var p1 = (data.FreePhysicalMemory / 1024 / 1024).toFixed(0), p2 = (data.TotalPhysicalMemory / 1024 / 1024).toFixed(0);
            var p3 = ((p2 - p1) / p2 * 100).toFixed(0);
            var pp = '<div class="progress" style="margin:5px 0"><div class="progress-bar progress-bar-warning" aria-valuenow="' + p3 + '" aria-valuemin="0" aria-valuemax="100" style="width: ' + p3 + '%;">' + p3 + '%</div></div>';
            htm.push('<td>' + pp + (p2 - p1) + ' / ' + p2 + ' MB</td>');
            htm.push('</tr>');

            if (data.SwapTotal) {
                htm.push('<tr>');
                htm.push('<td>Swap</td>');
                var p1 = (data.SwapFree / 1024 / 1024).toFixed(0), p2 = (data.SwapTotal / 1024 / 1024).toFixed(0);
                var p3 = ((p2 - p1) / p2 * 100).toFixed(0);
                var pp = '<div class="progress"><div class="progress-bar progress-bar-warning" aria-valuenow="' + p3 + '" aria-valuemin="0" aria-valuemax="100" style="width: ' + p3 + '%;">' + p3 + '%</div></div>';
                htm.push('<td>' + pp + (p2 - p1) + ' / ' + p2 + ' MB</td>');
                htm.push('</tr>');
            }

            htm.push('<tr>');
            htm.push('<td>磁盘</td>');
            htm.push('<td>');
            data.LogicalDisk.forEach(diskitem => {
                if (diskitem.Size > diskitem.FreeSpace && diskitem.Size >= 1024 * 1024) {
                    var dw = 'MB';
                    p1 = diskitem.Size - diskitem.FreeSpace;
                    p2 = diskitem.Size / 1024 / 1024;
                    p3 = (p1 / diskitem.Size * 100).toFixed(0);
                    p1 = p1 / 1024 / 1024;

                    if (diskitem.Size >= 1024 * 1024 * 1024) {
                        p1 = p1 / 1024;
                        p2 = p2 / 1024;
                        dw = 'GB';
                    }

                    pp = '<div class="progress" style="margin:5px 0"><div class="progress-bar progress-bar-warning" aria-valuenow="' + p3 + '" aria-valuemin="0" aria-valuemax="100" style="width: ' + p3 + '%;">' + p3 + '%</div></div>';
                    var dn = diskitem.Name || "";
                    dn = dn.length > 20 ? dn.substring(0, 20) + "..." : dn;
                    htm.push(pp + '<div>' + dn + ' &nbsp; ' + p1.toFixed(0) + ' / ' + p2.toFixed(0) + ' ' + dw + '</div>');
                }
            });
            htm.push('</td>');
            htm.push('</tr>');

            htm.push('<tr>');
            htm.push('<td>开机</td>');

            var p1 = new Date(data.TickCount - 8 * 1000 * 3600);
            var p2 = (p1.getFullYear() - 1970).toString();
            while (p2.length < 4) {
                p2 = "0" + p2;
            }
            var p3 = p1.getMonth() + 1 - 1;
            p3 < 10 && (p3 = "0" + p3);
            var p4 = p1.getDate() - 1;
            p4 < 10 && (p4 = "0" + p4);
            var p5 = p1.getHours();
            p5 < 10 && (p5 = "0" + p5);
            var p6 = p1.getMinutes();
            p6 < 10 && (p6 = "0" + p6);
            var p7 = p1.getSeconds();
            p7 < 10 && (p7 = "0" + p7);

            var pp = p2 + "-" + p3 + "-" + p4 + " " + p5 + ":" + p6 + ":" + p7;

            htm.push('<td style="letter-spacing:1px">' + pp + '</td>');
            htm.push('</tr>');

            htm.push('</table>');
            $('#divAs').html(htm.join(''));

            //自动刷新
            setTimeout(loadOSinfo, 1000 * 10);
        },
        error: function () {
            $('#divAs').html('<h4 class="text-center text-danger">获取服务器信息异常</h4>');
        }
    });
}

loadOSinfo();

src/Netnr.ResponseFramework/Netnr.Func/ExportAid.cs
using Netnr.Domain;
using NPOI.HSSF.UserModel;
using NPOI.SS.UserModel;
using NPOI.SS.Util;
using NPOI.XSSF.UserModel;
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;

namespace Netnr.Func
{
    /// <summary>
    /// 导出辅助
    /// </summary>
    public class ExportAid
    {
        /// <summary>
        /// 数据实体映射
        /// </summary>
        public static DataTable ModelsMapping(QueryDataInputVM ivm, QueryDataOutputVM ovm)
        {
            //转表（类型为字符串）
            DataTable dt = ovm.table;

            //更改列长度
            foreach (DataColumn col in dt.Columns)
            {
                col.MaxLength = short.MaxValue * 9;
            }

            var listColumns = ovm.columns as List<SysTableConfig>;

            //调整列排序
            var colorder = listColumns.Where(x => dt.Columns.Contains(x.ColField)).OrderBy(x => x.ColOrder).ToList();
            for (int i = 0; i < colorder.Count; i++)
            {
                var ci = colorder[i];
                if (dt.Columns.Contains(ci.ColField))
                {
                    dt.Columns[ci.ColField].SetOrdinal(i);
                }
            }

            #region 单元格转换
            foreach (DataRow dr in dt.Rows)
            {
                for (int i = 0; i < dt.Columns.Count; i++)
                {
                    string field = dt.Columns[i].ColumnName;
                    dr[i] = CellFormat(ivm.tableName, field, dr[i].ToString(), dr);
                }
            }
            #endregion

            //剔除不导出的列
            List<SysTableConfig> removeCol = listColumns.Where(x => x.ColExport != 1).ToList();
            foreach (SysTableConfig col in removeCol)
            {
                if (dt.Columns.Contains(col.ColField))
                {
                    dt.Columns.Remove(dt.Columns[col.ColField]);
                }
            }

            //剔除没在表配置的列
            List<string> removeColNotExists = new List<string>();
            foreach (DataColumn dc in dt.Columns)
            {
                if (listColumns.Where(x => x.ColField == dc.ColumnName).Count() == 0)
                {
                    removeColNotExists.Add(dc.ColumnName);
                }
            }
            foreach (string col in removeColNotExists)
            {
                dt.Columns.Remove(dt.Columns[col]);
            }

            //更改列名为中文（重复的列，后面追加4位随机数）
            foreach (SysTableConfig col in listColumns)
            {
                if (dt.Columns.Contains(col.ColField))
                {
                    try
                    {
                        dt.Columns[col.ColField].ColumnName = col.ColTitle;
                    }
                    catch (Exception)
                    {
                        dt.Columns[col.ColField].ColumnName = col.ColTitle + "-" + Core.RandomTo.NumCode();
                    }
                }
            }

            return dt;
        }

        /// <summary>
        /// 单元格格式化
        /// </summary>
        /// <param name="tableName">表名</param>
        /// <param name="field">字段名</param>
        /// <param name="value">单元格值</param>
        /// <param name="dr">当前行</param>
        /// <returns></returns>
        public static string CellFormat(string tableName, string field, string value, DataRow dr)
        {
            //格式化后的值
            string result = value;

            try
            {
                switch (tableName.ToLower())
                {
                    //角色
                    case "sysrole":
                        {
                            switch (field.ToLower())
                            {
                                //时间
                                case "srcreatetime":
                                    result = CellMapping.DateTimeFormat(value);
                                    break;

                                //状态
                                case "srstatus":
                                    result = CellMapping.Status01(value);
                                    break;
                            }
                        }
                        break;

                    //用户
                    case "sysuser":
                        {
                            switch (field.ToLower())
                            {
                                //角色ID
                                case "srid":
                                    result = dr["SrName"].ToString();
                                    break;

                                //时间
                                case "sucreatetime":
                                    result = CellMapping.DateTimeFormat(value, "yyyy-MM-dd HH:mm:ss");
                                    break;

                                //状态
                                case "sustatus":
                                    result = CellMapping.Status01(value);
                                    break;
                            }
                        }
                        break;

                    //日志
                    case "syslog":
                        {
                            switch (field.ToLower())
                            {
                                //时间
                                case "logcreatetime":
                                    result = CellMapping.DateTimeFormat(value, "yyyy-MM-dd HH:mm:ss");
                                    break;
                            }
                        }
                        break;

                    //数据字典
                    case "sysdictionary":
                        {
                            switch (field.ToLower())
                            {
                                //状态
                                case "sdstatus":
                                    result = CellMapping.Status03(value);
                                    break;
                            }
                        }
                        break;
                }
            }
            catch (Exception)
            {

            }

            return result;
        }

        /// <summary>
        /// 映射
        /// </summary>
        public class CellMapping
        {
            /// <summary>
            /// 缓存前缀
            /// </summary>
            public static string Mck = "CellMapping_";

            /// <summary>
            /// 缓存过期时间（单位：秒）
            /// </summary>
            public static int Mce = 300;

            /// <summary>
            /// 字典映射转换
            /// </summary>
            /// <param name="kv">格式：1:未生成,2:已生成</param>
            /// <param name="key"></param>
            /// <returns></returns>
            public static string KeyValueMap(string kv, string key)
            {
                var result = key;
                var kvs = kv.Split(',').ToList();
                foreach (var item in kvs)
                {
                    var ims = item.Split(':');
                    if (ims[0] == key)
                    {
                        result = ims[1];
                    }
                }

                return result;
            }

            /// <summary>
            /// 状态格式化
            /// </summary>
            /// <param name="value"></param>
            /// <returns></returns>
            public static string Status01(string value)
            {
                var kv = "1:✔,0:✘";
                return KeyValueMap(kv, value);
            }

            /// <summary>
            /// 状态格式化
            /// </summary>
            /// <param name="value"></param>
            /// <returns></returns>
            public static string Status02(string value)
            {
                var kv = "1:✘,0:✔";
                return KeyValueMap(kv, value);
            }

            /// <summary>
            /// 状态格式化
            /// </summary>
            /// <param name="value"></param>
            /// <returns></returns>
            public static string Status03(string value)
            {
                var kv = "-1:删除,1:正常,2:停用";
                return KeyValueMap(kv, value);
            }

            /// <summary>
            /// 时间格式化
            /// </summary>
            /// <param name="value"></param>
            /// <param name="format"></param>
            /// <returns></returns>
            public static string DateTimeFormat(string value, string format = "yyyy-MM-dd")
            {
                if (DateTime.TryParse(value, out DateTime dt))
                {
                    value = dt.ToString(format);
                }
                return value;
            }
        }

        /// <summary>
        /// 操作已经生成的Excel
        /// </summary>
        /// <param name="fullPath"></param>
        /// <param name="ivm"></param>
        /// <returns></returns>
        public static bool ExcelDraw(string fullPath, QueryDataInputVM ivm)
        {
            //需要绘制的记录
            var needDraw = "DatabaseTableDesign,syslog".ToLower().Split(',');
            if (!needDraw.Contains(ivm.tableName?.ToLower()))
            {
                return true;
            }

            string strExtName = Path.GetExtension(fullPath);

            IWorkbook workbook = null;

            using (FileStream file = new FileStream(fullPath, FileMode.OpenOrCreate, FileAccess.ReadWrite))
            {
                if (strExtName.Equals(".xls"))
                {
                    workbook = new HSSFWorkbook(file);
                }
                if (strExtName.Equals(".xlsx"))
                {
                    workbook = new XSSFWorkbook(file);
                }
            }

            ISheet sheet = workbook.GetSheetAt(workbook.ActiveSheetIndex);

            switch (ivm.tableName?.ToLower())
            {
                //数据库表设计
                case "databasetabledesign":
                    {
                        //冻结首行首列
                        sheet.CreateFreezePane(0, 1);

                        var rows = sheet.GetRowEnumerator();

                        var styleH = CreateCellStyle(workbook, StyleType.headLeft);

                        while (rows.MoveNext())
                        {
                            IRow row;
                            if (fullPath.Contains(".xlsx"))
                            {
                                row = (XSSFRow)rows.Current;
                            }
                            else
                            {
                                row = (HSSFRow)rows.Current;
                            }

                            var cc = row.GetCell(1);
                            if (string.IsNullOrWhiteSpace(cc.StringCellValue))
                            {
                                foreach (var cell in row.Cells)
                                {
                                    cell.CellStyle = styleH;
                                }

                                //合并
                                sheet.AddMergedRegion(new CellRangeAddress(row.RowNum, row.RowNum, 0, row.Cells.Count - 1));
                            }
                        }
                    }
                    break;
                default:
                    break;
            }

            using (FileStream file = new FileStream(fullPath, FileMode.OpenOrCreate))
            {
                workbook.Write(file);
                workbook.Close();
            }
            return false;
        }

        /// <summary>
        /// 单元格公共样式枚举
        /// </summary>
        public enum StyleType
        {
            /// <summary>
            /// 正常，默认
            /// </summary>
            normal,
            /// <summary>
            /// 头部居中
            /// </summary>
            headCenter,
            /// <summary>
            /// 头部居左
            /// </summary>
            headLeft,
            /// <summary>
            /// 白色字体，绿色背景
            /// </summary>
            whiteGreen,
            /// <summary>
            /// 白色字体，橙色背景
            /// </summary>
            whiteOrange,
            /// <summary>
            /// 蓝色字体
            /// </summary>
            blue,
            /// <summary>
            /// 绿色字体
            /// </summary>
            green,
            /// <summary>
            /// 黑色字体，灰色背景
            /// </summary>
            blankGray,
            /// <summary>
            /// 加粗
            /// </summary>
            bold,
            /// <summary>
            /// 头部无边框
            /// </summary>
            headNoBorder,
        }

        /// <summary>
        /// 定义单元格常用到样式
        /// </summary>
        /// <param name="wb"></param>
        /// <param name="st"></param>
        /// <returns></returns>
        static ICellStyle CreateCellStyle(IWorkbook wb, StyleType st = StyleType.normal)
        {
            ICellStyle cellStyle = wb.CreateCellStyle();

            //为避免日期格式被Excel自动替换，所以设定 format 为 『@』 表示一率当成text來看
            cellStyle.DataFormat = HSSFDataFormat.GetBuiltinFormat("@");
            cellStyle.BorderBottom = BorderStyle.Thin;
            cellStyle.BorderLeft = BorderStyle.Thin;
            cellStyle.BorderRight = BorderStyle.Thin;
            cellStyle.BorderTop = BorderStyle.Thin;
            //水平垂直居中
            cellStyle.Alignment = HorizontalAlignment.Center;
            cellStyle.VerticalAlignment = VerticalAlignment.Center;
            //背景颜色
            cellStyle.FillForegroundColor = 9;
            cellStyle.FillPattern = FillPattern.SolidForeground;

            //默认
            IFont font = wb.CreateFont();
            font.FontHeightInPoints = 10;
            font.Color = 8;

            switch (st)
            {
                case StyleType.normal:
                    break;
                case StyleType.headCenter:
                    font.FontHeightInPoints = 10;
                    font.IsBold = true;
                    break;
                case StyleType.headLeft:
                    font.FontHeightInPoints = 10;
                    font.IsBold = true;
                    cellStyle.Alignment = HorizontalAlignment.Left;
                    break;
                case StyleType.whiteGreen:
                    font.Color = 9;
                    cellStyle.FillForegroundColor = 17;
                    cellStyle.FillPattern = FillPattern.SolidForeground;
                    break;
                case StyleType.whiteOrange:
                    font.Color = 9;
                    cellStyle.FillForegroundColor = 51;
                    cellStyle.FillPattern = FillPattern.SolidForeground;
                    break;
                case StyleType.blue:
                    font.Color = 12;
                    break;
                case StyleType.green:
                    font.Color = 17;
                    break;
                case StyleType.blankGray:
                    font.Color = 8;
                    cellStyle.FillForegroundColor = 22;
                    cellStyle.FillPattern = FillPattern.SolidForeground;
                    break;
                case StyleType.bold:
                    font.IsBold = true;
                    break;
                case StyleType.headNoBorder:
                    font.FontHeightInPoints = 10;
                    font.IsBold = true;
                    break;
            }

            cellStyle.SetFont(font);

            return cellStyle;
        }
    }
}
